### DISCUSSION TITLE
Embedding in golang

### DISCUSSION THREAD

**Discussion by rajrahul** [cuelang/cue#992] [Q&A] [answered]
Planning to use cue for a document validation/transformation service. We will have a golang service that accept cue scripts, execute and return the resultant JSON.

The snippet below works as long as there are no library imports.

func evalCue(input string) (string, error) {
  res, err :=   rt.Compile("test.cue", input)
  if err != nil { ... }
  eval := res.Value().Eval()
  return convertCUEValueToString(eval)
}  
 
For the input below, looks like compilation removes the imports and results in error : "_|_ // str: cannot call non-function strings.ToUpper (type _|_) "

  body :=  `
   import "strings"	
   str : strings.ToUpper("hello world")
  `
   str, err := evalCue(body)
  
Any examples on how to handle this?

Thanks,
Rahul

**Comment by eonpatapon** [comment, marked as answer]
To use builtins you just need to import them in your go code like that:

```
import (
  _ "cuelang.org/go/pkg"
)
```

For managing imports of custom modules you will need to use `load.Instances` instead of `Runtime.Compile`.

**Reply by rajrahul** [reply]
Thanks it is working now. Is there way to pass in memory strings than to read from files on disk?

**Reply by verdverm** [reply]
I believe the plan is to provide a fs.FS option to the load config once Go 1.17 is available (really once Go 1.15 is no longer supported)

You may be able to use the Overlay option, can't recall for sure without testing.

**Reply by myitcv** [reply]
https://github.com/cuelang/cue/issues/607#issuecomment-797878959 indeed remains the plan. 

@rajrahul depending on your use case you  might be able to rely on the `Overlay` provided as part of https://pkg.go.dev/cuelang.org/go@v0.3.2/cue/load#Config. Otherwise, you might need to use `go:embed` directive (or equivalent), write to a temp directory, and load from there.

**Comment by cueckoo** [comment]
This discussion has been migrated to https://github.com/cue-lang/cue/issues/992.

For more details about CUE's migration to a new home, please see https://github.com/cue-lang/cue/issues/1078.
